### DISCUSSION TITLE
Have RefreshDatabase apply between each test function instead of each test class to reset database state between tests

### DISCUSSION THREAD

**Discussion by drewg2009** [laravel/framework#46387] [Ideas]
Right now the documentation on RefreshDatabase is not clear that it doesn't refresh between each individual test function and that it only resets the database between each test class. 

There also doesn't seem to be any examples of the other traits that clear and migrate the database. 

It would be so nice if each test function could start with a new database state so they are completely isolated between runs

**Comment by henzeb** [comment]
If documentation were to be changed. I think it should add a paragraph where it explains how to make it refresh the database between each test. 

```php
use RefreshDatabase;

public function setUp(): void
    {
        parent::setUp();

        $this->refreshDatabase();
    }

```

**Reply by drewg2009** [reply]
After more testing it seems like it does refresh between test functions based on the database engine you are using. 

One developer was using a MyISAM engine for his db in php my admin, which caused it to persist the data since that engine is not transactional. 

I think that the docs need to be expanded quite a bit to cover for these cases and to explain what each trait does

**Reply by humanatepixel** [reply]
Is there a chance someone here knows how to do the opposite? I'd like the data to persist between two methods of the same class that uses the `RefreshDatabase` trait

**Reply by AJenbo** [reply]
@drewg2009 it uses transactions to reset the database to the same state as before the test. Make sure what you are doing works with transactions in the database you use, or use DatabaseMigrations instead.

@humanatepixel that's not something you can or should do.
